src/dotnet/GeneratorAgent/src/Azure.Tools.GeneratorAgent/GitHubFileService.cs

```diff
@@ -0,0 +1,158 @@
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Azure.Tools.GeneratorAgent.Configuration;
+using Microsoft.Extensions.Logging;
+
+namespace Azure.Tools.GeneratorAgent
+{
+    /// <summary>
+    /// TypeSpec file service for GitHub repository access using GitHub API for fast file retrieval.
+    /// </summary>
+    internal class GitHubFileService
+    {
+        private readonly ILogger<GitHubFileService> Logger;
+        private readonly AppSettings AppSettings;
+        private readonly HttpClient HttpClient;
+        private readonly string CommitId;
+        private readonly string TypespecSpecDir;
+
+        // Static JSON options for better performance (reused across calls)
+        private static readonly JsonSerializerOptions JsonOptions = new()
+        {
+            PropertyNameCaseInsensitive = false,
+            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull
+        };
+
+        public GitHubFileService(
+            AppSettings appSettings,
+            ILogger<GitHubFileService> logger,
+            ValidationContext validationContext,
+            HttpClient httpClient)
+        {
+            ArgumentNullException.ThrowIfNull(appSettings);
+            ArgumentNullException.ThrowIfNull(logger);
+            ArgumentNullException.ThrowIfNull(validationContext);
+            ArgumentNullException.ThrowIfNull(httpClient);
+
+            AppSettings = appSettings;
+            Logger = logger;
+
+            CommitId = validationContext.ValidatedCommitId;
+            TypespecSpecDir = validationContext.ValidatedTypeSpecDir;
+
+            HttpClient = httpClient;
+            
+            // Log GitHub authentication status
+            bool hasAuthHeader = httpClient.DefaultRequestHeaders.Authorization != null;
+            Logger.LogInformation("GitHub API client initialized. Authentication: {AuthStatus}", 
+                hasAuthHeader ? "Configured" : "Not configured - using rate-limited access");
+        }
+
+        public async Task<Result<Dictionary<string, string>>> GetTypeSpecFilesAsync(CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                Result<Dictionary<string, string>> apiResult = await FetchTypeSpecFilesViaApiAsync(cancellationToken).ConfigureAwait(false);
+                return apiResult;
+            }
+            catch (Exception ex) when (ex is not OperationCanceledException)
+            {
+                Logger.LogCritical(ex, "Unexpected error while fetching TypeSpec files from GitHub repository {Repository}", 
+                    AppSettings.AzureSpecRepository);
+                throw;
+            }
+        }
+
+        private async Task<Result<Dictionary<string, string>>> FetchTypeSpecFilesViaApiAsync(CancellationToken cancellationToken)
```

Comment:
I feel like the error handling in this method is a bit inconsistent. i.e. 
- My get directory contents fails -> immediate and specific failure
- All of my DownloadFileContentAsync tasks threw an exception -> success and return an empty dictionary